### PR TITLE
DOP-4214: Compose list of directives used per page

### PR DESCRIPTION
**Pull request by seungpark**
### Stories/Links:

DOP-4214

This change will create a new graphql node `PageDirective` that will hold information about the page's directives and node types used. A page query has been added so that each page is aware of the node types and directive types present in the AST. In doing so, had to convert the [in-memory page creation in our build process](https://github.com/mongodb/snooty/pull/991/files#diff-446f3d9f3161103ad633d9ecd573872029f2df7b7e38e005b6da39cad2361f14L16-L21) to one that is more adhere to Gatsby lifecycles:
- During `sourceNodes` API, we create nodes for each page that is sourced from manifest zip file
- During `createPages` API, we run a [graphql query](https://github.com/mongodb/snooty/pull/991/files#diff-446f3d9f3161103ad633d9ecd573872029f2df7b7e38e005b6da39cad2361f14R254-R264) to refer to these pages, and only pass the `page_id` to the PageContext
- Each page component (in this case DocumentBody) should have an [exported page query that returns the AST](https://github.com/mongodb/snooty/pull/991/files#diff-35aa43db167c445bb5545b721164693310ea74f1c394aa9bbd55148967c03206R209-R215)

### Current Behavior:

No difference in behavior of page. We are constructing a list of directives and node types used in the page, which will be used for consumption by [DOP-4215](https://jira.mongodb.org/browse/DOP-4215).

[master branch staged](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/docs/docsworker-xlarge/master/index.html)

### Staging Links (no functional change. staged to verify that there is no front end performance degradation):

https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/docs/docsworker-xlarge/DOP-4214/index.html

### Build logs to show time to build front end:
[Docs for front end build (on branch) reads Done building in 255.736645165 sec](https://workerpoolstaging-qgeyp.mongodbstitch.com/pages/job.html?collName=queue&jobId=65bc145f6b2f17c95addeb94)
[front end build 2 reads ](https://workerpoolstaging-qgeyp.mongodbstitch.com/pages/job.html?collName=queue&jobId=65b7cb5594771f7657c2e654)
[Cloud docs build time for front end reads Done building in 184.534027831 sec](https://workerpoolstaging-qgeyp.mongodbstitch.com/pages/job.html?collName=queue&jobId=65bc18ef6b2f17c95adfc80b)

Sample logs from outside of branch in same collection (pool_test.queue):
[cloud docs](https://workerpoolstaging-qgeyp.mongodbstitch.com/pages/job.html?collName=queue&jobId=65bc11326b2f17c95adba9ed)
[docs](https://workerpoolstaging-qgeyp.mongodbstitch.com/pages/job.html?collName=queue&jobId=65bae0616b2f17c95a7debac)

### CODE REVIEW

**Comment by seungpark** [comment]
After discussion with @casthewiz - will update this PR to have the query be done on createPages step of gatsby build steps, so that this can be available at this step. Upon first glance, it looks like a good idea to move the in-memory page data to graphql as well, as we cannot have interpolated strings in graphql queries, but rather iterate over the results as a list

File: plugins/gatsby-source-snooty-prod/gatsby-node.js [hunk 1]
```diff
@@ -247,26 +250,41 @@ exports.createPages = async ({ actions }) => {
...
 
-      if (RESOLVED_REF_DOC_MAPPING[page] && Object.keys(RESOLVED_REF_DOC_MAPPING[page]).length > 0) {
-        createPage({
-          path: assertTrailingSlash(slug),
-          component: path.resolve(__dirname, mainComponentRelativePath),
-          context: {
-            slug,
-            repoBranches,
-            template: pageNodes?.options?.template,
-            page: pageNodes,
```
[the remark this answers is not in the document]
Comment (seungpark) [reply plugins/gatsby-source-snooty-prod/gatsby-node.js:285]: added the graphql to the layout component!

File: plugins/gatsby-source-snooty-prod/gatsby-node.js [hunk 2]
```diff
@@ -247,26 +250,41 @@ exports.createPages = async ({ actions }) => {
...
   }
 
+  // DOP-4214: for each page, query the directive/node types
+  const pageList = await graphql(`
+    {
+      allPage {
+        nodes {
+          page_id
+          ast
+          componentNames
```
[the remark this answers is not in the document]
Comment (seungpark) [reply plugins/gatsby-source-snooty-prod/gatsby-node.js:278]: this is not being used currently, but it is intended to be used along [these lines](https://github.com/mongodb/snooty/pull/991/files#diff-446f3d9f3161103ad633d9ecd573872029f2df7b7e38e005b6da39cad2361f14R289-R304) to create page of only necessary components

File: plugins/gatsby-source-snooty-preview/gatsby-node.js [hunk 3]
```diff
@@ -54,6 +55,12 @@ exports.createSchemaCustomization = async ({ actions }) => {
     type AssociatedProduct implements Node @dontInfer {
       productName: String
     }
+
+    type PageDirective implements Node @dontInfer {
```
[the remark this answers is not in the document]
Comment (seungpark) [reply plugins/gatsby-source-snooty-preview/gatsby-node.js]: good catch! removed

File: src/layouts/index.js [hunk 4]
```diff
@@ -145,3 +149,11 @@ DefaultLayout.propTypes = {
 };
 
 export default DefaultLayout;
+
+export const query = graphql`
```
Comment (rayangler) [inline src/layouts/index.js:153]: I don't think this `query` is actually being invoked. The data passed to the layout component is most likely from the page query for `DocumentBody`. Data is still passed when this is commented out. We can probably remove this

File: plugins/gatsby-source-snooty-preview/gatsby-node.js [hunk 5]
```diff
@@ -234,11 +235,13 @@ exports.createPages = async ({ actions, createNodeId, getNode, graphql, reporter
         );
         return;
       }
+      console.log(`createPage with id ${node.pageNodeId}`);
```
Comment (rayangler) [inline plugins/gatsby-source-snooty-preview/gatsby-node.js:238]: Should we remove this?

File: src/layouts/index.js (hunk 4 above)
Comment (seungpark) [reply src/layouts/index.js]: good catch! tested and verified. pushed the change!

**Review by rayangler — APPROVED** [review]

**Pull request merged**